File: samples/ModelBatch/src/ModelBatchSample.cpp
// Oak Nut include
#include <onut/Anim.h>
#include <onut/Model.h>
#include <onut/Renderer.h>
#include <onut/Settings.h>
#include <onut/Shader.h>
#include <onut/Texture.h>
#include <onut/Timing.h>

OModelRef pScene;

void initSettings()
{
    oSettings->setGameName("Model Batch Sample");
    oSettings->setShowFPS(true);
}

void init()
{
    oRenderer->setSun(Vector3::UnitZ, Color::Black);
    oRenderer->setAmbient(Color(0.05f, 0.075f, 0.1f));
    oRenderer->setLight(0, Vector3(1, 5.0f, 1.5f), 3.0f, Color(0.5f, 1.0f, 1.5f)); // Stairs
    oRenderer->setLight(1, Vector3(1, 0, 1.1f), 1.3f, Color(0.5f, 1.0f, 1.5f)); // Light well
    oRenderer->setLight(2, Vector3(1, 2, 0.5f), 1.5f, Color(1.5f, 1.0f, 0.5f)); // Torch

    pScene = OModel::createFromBatch({
        { OGetModel("floor.model"), Matrix::CreateTranslation({0, 0, 0}) },
        { OGetModel("floor.model"), Matrix::CreateTranslation({1, 0, 0}) },
        { OGetModel("floor.model"), Matrix::CreateTranslation({2, 0, 0}) },
        { OGetModel("floor.model"), Matrix::CreateTranslation({0, 1, 0}) },
        { OGetModel("floor.model"), Matrix::CreateTranslation({1, 1, 0}) },
        { OGetModel("floor.model"), Matrix::CreateTranslation({2, 1, 0}) },
        { OGetModel("floor.model"), Matrix::CreateTranslation({0, 2, 0}) },
        { OGetModel("floor.model"), Matrix::CreateTranslation({1, 2, 0}) },
        { OGetModel("ceiling.model"), Matrix::CreateTranslation({0, 0, 0}) },
        { OGetModel("lightWell.model"), Matrix::CreateTranslation({1, 0, 0}) },
        { OGetModel("ceiling.model"), Matrix::CreateTranslation({2, 0, 0}) },
        { OGetModel("ceiling.model"), Matrix::CreateTranslation({0, 1, 0}) },
        { OGetModel("ceiling.model"), Matrix::CreateTranslation({1, 1, 0}) },
        { OGetModel("ceiling.model"), Matrix::CreateTranslation({2, 1, 0}) },
        { OGetModel("ceiling.model"), Matrix::CreateTranslation({0, 2, 0}) },
        { OGetModel("ceiling.model"), Matrix::CreateTranslation({1, 2, 0}) },
        { OGetModel("stairs.model"), Matrix::CreateTranslation({.5f, 2.5f, 0}) },
        { OGetModel("wall.model"), Matrix::CreateTranslation({-.5f, 2.5f, 0}) },
        { OGetModel("wall.model"), Matrix::CreateRotationZ(-OPIDIV2) * Matrix::CreateTranslation({1.5f, 2.5f, 0}) },
        { OGetModel("wall.model"), Matrix::CreateTranslation({1.5f, 1.5f, 0}) },
        { OGetModel("wall.model"), Matrix::CreateRotationZ(-OPIDIV2) * Matrix::CreateTranslation({2.5f, 1.5f, 0}) },
        { OGetModel("wall.model"), Matrix::CreateRotationZ(-OPIDIV2) * Matrix::CreateTranslation({2.5f, 0.5f, 0}) },
        { OGetModel("column.model"), Matrix::CreateTranslation({-.5f, 2.5f, 0}) },
        { OGetModel("column.model"), Matrix::CreateTranslation({.5f, 2.5f, 0}) },
        { OGetModel("column.model"), Matrix::CreateTranslation({1.5f, 2.5f, 0}) },
        { OGetModel("column.model"), Matrix::CreateTranslation({1.5f, 1.5f, 0}) },
        { OGetModel("column.model"), Matrix::CreateTranslation({2.5f, 1.5f, 0}) },
        { OGetModel("column.model"), Matrix::CreateTranslation({2.5f, 0.5f, 0}) },
        { OGetModel("column.model"), Matrix::CreateTranslation({2.5f, -.5f, 0}) }
    });
}

void update()
{
}

void render()
{
    oRenderer->setupFor3D({0, 0, 0.5f}, {1, 1, 0.5f}, Vector3::Up, 90);
    oRenderer->clear({0, 0, 0, 1});
    oRenderer->clearDepth();

    pScene->render(Matrix::Identity);
}

void postRender()
{
}

void renderUI()
{
}

File: samplesJS/ModelBatch/assets/scripts/ModelBatchSample.js
var scene = Model.createFromBatch([
    { model: getModel("floor.model"), transform: Matrix.createTranslation(new Vector3(0, 0, 0)) },
    { model: getModel("floor.model"), transform: Matrix.createTranslation(new Vector3(1, 0, 0)) },
    { model: getModel("floor.model"), transform: Matrix.createTranslation(new Vector3(2, 0, 0)) },
    { model: getModel("floor.model"), transform: Matrix.createTranslation(new Vector3(0, 1, 0)) },
    { model: getModel("floor.model"), transform: Matrix.createTranslation(new Vector3(1, 1, 0)) },
    { model: getModel("floor.model"), transform: Matrix.createTranslation(new Vector3(2, 1, 0)) },
    { model: getModel("floor.model"), transform: Matrix.createTranslation(new Vector3(0, 2, 0)) },
    { model: getModel("floor.model"), transform: Matrix.createTranslation(new Vector3(1, 2, 0)) },
    { model: getModel("ceiling.model"), transform: Matrix.createTranslation(new Vector3(0, 0, 0)) },
    { model: getModel("lightWell.model"), transform: Matrix.createTranslation(new Vector3(1, 0, 0)) },
    { model: getModel("ceiling.model"), transform: Matrix.createTranslation(new Vector3(2, 0, 0)) },
    { model: getModel("ceiling.model"), transform: Matrix.createTranslation(new Vector3(0, 1, 0)) },
    { model: getModel("ceiling.model"), transform: Matrix.createTranslation(new Vector3(1, 1, 0)) },
    { model: getModel("ceiling.model"), transform: Matrix.createTranslation(new Vector3(2, 1, 0)) },
    { model: getModel("ceiling.model"), transform: Matrix.createTranslation(new Vector3(0, 2, 0)) },
    { model: getModel("ceiling.model"), transform: Matrix.createTranslation(new Vector3(1, 2, 0)) },
    { model: getModel("stairs.model"), transform: Matrix.createTranslation(new Vector3(.5, 2.5, 0)) },
    { model: getModel("wall.model"), transform: Matrix.createTranslation(new Vector3(-.5, 2.5, 0)) },
    { model: getModel("wall.model"), transform: Matrix.createRotationZ(-90).mul(Matrix.createTranslation(new Vector3(1.5, 2.5, 0))) },
    { model: getModel("wall.model"), transform: Matrix.createTranslation(new Vector3(1.5, 1.5, 0)) },
    { model: getModel("wall.model"), transform: Matrix.createRotationZ(-90).mul(Matrix.createTranslation(new Vector3(2.5, 1.5, 0))) },
    { model: getModel("wall.model"), transform: Matrix.createRotationZ(-90).mul(Matrix.createTranslation(new Vector3(2.5, 0.5, 0))) },
    { model: getModel("column.model"), transform: Matrix.createTranslation(new Vector3(-.5, 2.5, 0)) },
    { model: getModel("column.model"), transform: Matrix.createTranslation(new Vector3(.5, 2.5, 0)) },
    { model: getModel("column.model"), transform: Matrix.createTranslation(new Vector3(1.5, 2.5, 0)) },
    { model: getModel("column.model"), transform: Matrix.createTranslation(new Vector3(1.5, 1.5, 0)) },
    { model: getModel("column.model"), transform: Matrix.createTranslation(new Vector3(2.5, 1.5, 0)) },
    { model: getModel("column.model"), transform: Matrix.createTranslation(new Vector3(2.5, 0.5, 0)) },
    { model: getModel("column.model"), transform: Matrix.createTranslation(new Vector3(2.5, -.5, 0)) }
])

Renderer.setSun(Vector3.UNIT_Z, Color.BLACK);
Renderer.setAmbient(new Color(0.05, 0.075, 0.1));
Renderer.setLight(0, new Vector3(1, 5.0, 1.5), 3.0, new Color(0.5, 1.0, 1.5)); // Stairs
Renderer.setLight(1, new Vector3(1, 0, 1.1), 1.3, new Color(0.5, 1.0, 1.5)); // Light well
Renderer.setLight(2, new Vector3(1, 2, 0.5), 1.5, new Color(1.5, 1.0, 0.5)); // Torch

function render()
{
    Renderer.setupFor3D(
        new Vector3(0, 0, 0.5),
        new Vector3(1, 1, 0.5),
        Vector3.UNIT_Z,
        90
    )

    Renderer.clear(new Color(0, 0, 0, 1))
    Renderer.clearDepth();

    scene.render(Matrix.IDENTITY);
}
